Firestore/core/src/util/executor_libdispatch.mm

```diff
@@ -243,91 +113,184 @@ void MarkDone() {
-  TimeSlot* time_slot = nullptr;
-  TimeSlotId time_slot_id = 0;
-  RunSynchronized(this, [this, delay, &operation, &time_slot, &time_slot_id] {
-    time_slot_id = NextId();
-    time_slot = new TimeSlot{this, delay, std::move(operation), time_slot_id};
-    schedule_[time_slot_id] = time_slot;
-  });
-
-  dispatch_after_f(delay_ns, dispatch_queue(), time_slot,
-                   TimeSlot::InvokedByLibdispatch);
-
-  return DelayedOperation{[this, time_slot_id] {
+  // to truly cancel work after it's been dispatched, libdispatch is guaranteed
+  // to outlive the executor, and it's possible for work to be invoked by
+  // libdispatch after the executor is destroyed. The Executor only stores an
+  // observer pointer to the operation.
+  Task* task = nullptr;
+  TimePoint target_time = MakeTargetTime(delay);
+  Id id = 0;
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    id = NextIdLocked();
+    task = Task::Create(this, target_time, tag, id, std::move(operation));
+    async_tasks_.insert(task);
+    schedule_[id] = task;
+  }
+
+  dispatch_after_f(delay_ns, dispatch_queue_, task, InvokeAsync);
+
+  return DelayedOperation(this, id);
+}
+
+void ExecutorLibdispatch::Complete(Task* task) {
+  bool should_release = false;
+  {
+    TASK_TRACE("Executor::Complete %s task %s", this, task);
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto found = async_tasks_.find(task);
+    if (found != async_tasks_.end()) {
+      should_release = true;
+      async_tasks_.erase(found);
+
+      if (!task->is_immediate()) {
+        schedule_.erase(task->id());
+      }
+    }
+  }
+
+  if (should_release) {
+    task->Release();
+  }
+}
+
+void ExecutorLibdispatch::Cancel(Id operation_id) {
+  Task* found_task = nullptr;
+  {
+    TASK_TRACE("Executor::Cancel %s task %s", this, task);
+    std::lock_guard<std::mutex> lock(mutex_);
+
     // `time_slot` might have been destroyed by the time cancellation function
     // runs, in which case it's guaranteed to have been removed from the
     // `schedule_`. If the `time_slot_id` refers to a slot that has been
     // removed, the call to `RemoveFromSchedule` will be a no-op.
-    RemoveFromSchedule(time_slot_id);
-  }};
-}
-
-void ExecutorLibdispatch::RemoveFromSchedule(TimeSlotId to_remove) {
-  RunSynchronized(this, [this, to_remove] {
-    const auto found = schedule_.find(to_remove);
+    const auto found = schedule_.find(operation_id);
 
     // It's possible for the operation to be missing if libdispatch gets to run
     // it after it was force-run, for example.
     if (found != schedule_.end()) {
-      found->second->MarkDone();
+      found_task = found->second;
+      async_tasks_.erase(found_task);
       schedule_.erase(found);
     }
-  });
+  }
+
+  if (found_task) {
+    found_task->Cancel();
+    found_task->Release();
+  }
 }
 
-// Test-only methods
+void ExecutorLibdispatch::InvokeAsync(void* raw_task) {
+  auto task = static_cast<Task*>(raw_task);
```

Comment:
Nit: consider qualifying `auto` with an asterisk to make it (even) clearer that it's a pointer.